Rewrite JAR loader

Simn opened the pull request:
This is still a bit work-in-progress. I'll enable it by default for the JVM target and add a flag to use the old loader. ~For now it can be tested with `-D java-modern-loader`.~ It is enabled by default and can be switched off with `-D jar-legacy-loader`.

There's a whole bunch of code in `java.ml` that does "something". It's very hard for me to tell how much of that is only there because genjava generates syntax. We will probably come across various regressions in various .jar libs and then can work towards fixing those.

closes #3397
closes #5725
closes #9569
closes #9570
closes #9573

Edit: I should probably mention that I'm only adapting this to support the JVM target, not the old Java target.

Simn committed:
[jvm] start writing a new jar-loader because I hate myself

Simn committed:
[jvm] parse MethodParameters attribute

closes #9573

Simn committed:
[jvm] try something completely different

Simn committed:
[jvm] detect self-constrained type parameters

because of course scala has that

Simn committed:
[jvm] ignore private types when loading jars

Simn committed:
[jvm] clean up loader code a bit

Simn committed:
[jvm] try treating trailing dollars differently

Simn committed:
[jvm] enable by default, use -D java-legacy-loader to disable

Gama11 commented:
Perhaps it should be `jvm-legacy-loader` or `jar-legacy-loader` if it doesn't actually work on the Java target?

Simn committed:
[jvm] add some timers

Simn committed:
[jvm] fix file listing

Simn committed:
[jvm] fix random _Statics occurrence

Simn committed:
[jvm] rename to -D jar-legacy-loader

Simn committed:
[jvm] try a custom reader

Simn committed:
[jvm] clean up a bit

Simn committed:
[jvm] List.init non c'e

Gama11 committed:
[typer] improve the error message for static fields in interfaces

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
[jvm] fix non-static local variable table lookups

Simn committed:
[typer] don't actually include extra arguments in the call argument list

This confuses overload resolution. It should be enough to type the things in case we hover over them or something.

Simn committed:
[jvm] more safety

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
[jvm] bring back annotation parsing and AOverride detection

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
Merge branch 'development' into java-loader-modern

Simn committed:
[jvm] ExtList
